Feature/korg kaoss dj

larromba opened the pull request:
I've been working to get the Korg Kaoss DJ working with Mixxx 2.1. In this pull request:
* Default Korg KAOSS Controller Mapping (xml / js)
* MIDI OUT bug fix for Mac 10.12 & Korg  Kaoss DJ 1.6

I initially read this forum thread:
https://www.mixxx.org/forums/viewtopic.php?f=7&t=8479&start=20

It took a few days but I noticed the midi output didn't seem to be working with the DJ unit (i.e. no LEDs were being lit up). It seems to work for people in the thread, so I assumed they might be using Windows or Linux. I eventually downloaded MIDI monitor and also ran Mixxx in `--debug` mode. It seemed no MIDI messages were being sent to the unit. I then built the source and found the MIDI IN and MIDI OUT weren't being linked up in `portmidienumerator.cpp` because of an edge case naming issue. I've had a go at fixing it and it now works as expected.

I prefer Mixxx over Serato DJ Intro, as it doesn't crash so much & supports more audio types - so would be great to get this accepted soon into 2.1 or 2.2. Hopefully it'll help other people out who may have this cool little Korg Kaoss thing also.

larromba committed:
+ Korg Kaoss DJ XML / JS mappings

larromba committed:
fix: MIDI OUT bug with Mac OS 10.12 & Korg Kaoss DJ 1.6

Be-ing commented:
Thanks for submitting this and fixing the issue with MIDI output on macOS! It will probably be a few weeks before I get to review this thoroughly because there is a big backlog of mappings to review and I am currently focusing on the manual for 2.1. To be clear, [this wiki page](https://mixxx.org/wiki/doku.php/korg_kaoss_dj) documents this mapping, right?

larromba committed:
fix: beatgrid move bug

larromba commented:
@Be-ing no worries. ok sure - for now I'll continue with building from source. the wiki isn't quite up to date with this mapping yet. when would be a good time to update it?

Be-ing commented:
Update the wiki whenever you get a chance.

naught101 commented:
Wow. I did not know about that controller, but now I REALLY want it. Have been looking for something minimal and light-weight.

How complete is the mapping? Are there any features of the controller that you can't use? (e.g. FX?)

uklotzde commented:
Nice. I have one and will test it.

larromba commented:
@Be-ing ok will do. I will aim for early next week. @naught101 i got it because i wanted to learn to DJ, but wanted something small. I was hoping I could also use the controller in music production too. The FX stuff is built in, and affects the master channel only, but the fx are cool and mostly intuitive. Hopefully it will give some interesting options whilst mixing. I'm a fan of it @uklotzde great - let us know how it goes

Be-ing commented:
Generally, following the labels on the device is good, but the mapping making sense with how Mixxx works is more important than exactly copying how it works with other software.

larromba commented:
@Be-ing ok good to know. I think it's very similar to how it works with Serato Intro. The xml / js mapping code is hacked up from the mapping on the forum post. I think all major elements are working (jogs, loop, faders, all buttons / leds). Maybe all that's left is finding a shift operation for backwards / fowards beat skip. One thing I noticed is that you can't elongate or shorten a loop if you explictly pick the 'in' and 'out' points. It works when the loop is 'quantised'. Is that expected?

Be-ing commented:
The loop_in and loop_out Controls need to be set to 1 on button press and 0 on button release. If both of those pieces are not mapped correctly you may get strange behavior.

fayaaz commented:
If this is hacked from my script (and it looks like it is :) ) the FX are not working and neither is beatlooproll. I'll have a look at elongating or shortening a manual loop.

larromba commented:
@fayaaz Yes it's all hacked from the script - are the changes cool with you? I forked your script on github if that helps. Use Shift + Middle strip button (when in loop mode) to start a quantised loop. The Middle-Strip button (without shift) then turns the quantised loop on / off. I feel there wasn't so much a need for the JS code in your script for the looping as it all seemed mappable using the built in MIDI-learn. What do you think? Does it work for you if you use the Shift + middle strip button (when in loop mode)?

larromba commented:
I've updated the [wiki](https://www.mixxx.org/wiki/doku.php/korg_kaoss_dj)

In doing so, I identified these as potential TODOs (also documented in the wiki):

- Browseknob + Shift --> Moves between folder levels
- Touchpad (controller mode) --> Controls the effects of the DJ software
- Touchpad (controller mode) + Shift --> Adjusts the Beats Multiplier
- Touchpad (sampler mode) --> Controls the sampler function of the DJ software
- Confusion around Loop x1, Loop on/off & Loop roll mode --> So looping mostly works, but as @fayaaz pointed out the loop roll functionality isn't implemented. @fayaaz looks like you might be right using JS here - I guess it's needed to handle different actions for the Loop x1 button?
- FX button - Applies the DJ software effect to the selected deck

Currently the controller is usable without these TODOs. Maybe the Mixxx software FX aren't 100% neccessary as the controller comes with a built-in KAOSS pad (but only for the master channel). The looping should probably be improved - I guess loop roll is an important feature. @Be-ing any thoughts?

larromba committed:
added Fx functionality

When Fx button left is enabled, touchpad can be used to change FX1 rack
mix volume.

When Fx button right is enabled, touchpad can be used to change FX2
rack mix volume.

larromba committed:
Fx rack mix set to 0 on touch up (mimicking serato dj intro)

bgmnt commented:
Hey, since the discussion mainly moved here, i want to repeat my question:

Is there a way, to have a different bevahiour to the jog wheels, than there is at the moment? Like: When a song is playing, "turning" the jog wheel is like going forward/backward in the track (like in a tape deck, while it plays). By hitting Shift plus "turning" the jog wheel a fast forward/backward is possible, like when stopping a tape deck and then going fwd/bwd. The difference to the existing shift+touchstrip strip search: The position in the track follows the move of the jog wheel relatively. Since I never scratch, I used the jog wheel heavily to navigate in tracks on the non-playing deck - and Virtual DJ e.g. hat the desired behaviour on by default. It should not be so exotic then :]
Perhaps, when someone points me to the nescessary function names, I could play for myself with it, but I am not very optimistic with this.

greetings!

larromba commented:
@bgmnt unfortunately I'm not sure if I understand your question fully. I know this is the functionality at the moment:

- **Scratch turned on**: jog wheel moves forward/backward in track 
- **Scratch turned on + shift**: jog wheel moves forward/backward in track (fast) 
- **Scratch turned off**: jog wheel speeds up / slows down the track 
- **Scratch turned off + shift**: jog wheel moves beatgrid
- **Touchstrip + shift**: moves playhead to exact point in the track

Are you asking, is there a way to make the jog wheels more like the touchstrip?

bgmnt commented:
@larromba I think, you got me right. Although I don't know, how touchstrips in general work, since i never really used it before. In @fayaaz fayaaz' mapping it worked like pointing somewhere in the song, but moving your finger did not move the waveform.

Anyway - what you described in bold letters, is exactly what I mean and what I had in Virtual DJs mapping of the device. I assume, the Serato Intro Mapping behaviour is the same? In fayaaz' mapping I didn't have this working (scratch mode only), but perhaps it was and I have another problem: I don't know, how to switch scratch mode one and off. No button in Mixxx and on the controller? Sorry, i think I'm stuck a bit here :)

larromba commented:
@bgmnt the touch trips still work like that, you just need to hold shift and then touch. First make sure the strip is in "Normal" mode. You change the mode by pressing the "Touch Slider mode button". To turn on / off scratch mode,  press the middle strip button (the red one) when in "Normal" mode. Refer to the [wiki](https://www.mixxx.org/wiki/doku.php/korg_kaoss_dj) for more help.

All I know is the stuff in bold above works in [this](https://github.com/larromba/mixxx_kaoss_dj) mapping. Perhaps try it if you havn't already?

Does that help you?

fayaaz commented:
@bgmnt I made some updates to the jog wheels and I believe from what @larromba is saying he also has made some updates.

> @fayaaz Yes it's all hacked from the script - are the changes cool with you?

Of course, this is the nature of open source :)

fayaaz commented:
> In @fayaaz fayaaz' mapping it worked like pointing somewhere in the song, but moving your finger did not move the waveform.

@bgmnt I don't think the korg sends midi messages when moving your finger, which is why you are seeing this behaviour

bgmnt commented:
@fayaaz @larromba Thank you guys, sounds great! I think I did not get your progress in detail since you move the developement to here and develop that fast :D I definitely have to test it, yes (in a few days i can do it).
The behaviour of the touch strip is okay for me, since it works good together with a 2-speed-jog-turning behaviour.

I really like, what happens here :+1:

Be-ing commented on res/controllers/Korg-KAOSS-DJ-scripts.js:
```diff
@@ -0,0 +1,194 @@
+    }
+};
+
+KAOSSDJ.controllerFxTouchMove = function(channel, control, value, status, group) {
```
Does the controller send MIDI messages just for one axis of the touchpad? I would presume that it would send one MIDI message for the horizontal axis and one for the vertical axis so you could control two things at once. If it does that, you could have one axis control the dry/wet knob (`mix`) and the other control the superknob (`super1`).

Be-ing commented on res/controllers/Korg-KAOSS-DJ-scripts.js:
```diff
@@ -0,0 +1,194 @@
+
+function KAOSSDJ() {};
+                     
+const ON = 0x7F, 
```
I am surprised that `const` works with Qt4's ancient JS interpreter. My guess is that the interpreter ignores the `const`, Since the interpreter doesn't actually support this keyword, please change this to `var` to avoid confusion.

Be-ing commented on res/controllers/Korg-KAOSS-DJ.midi.xml:
```diff
@@ -0,0 +1,902 @@
+            </control>
+            <control>
+                <group>[Sampler3]</group>
+                <key>cue_cdj</key>
```
Use `cue_gotoandplay` for samplers. This allows for repeatedly pressing the button to retrigger the sample before it finishes playing.

Be-ing commented on res/controllers/Korg-KAOSS-DJ-scripts.js:
```diff
@@ -0,0 +1,194 @@
+// Korg KAOSS DJ controller mapping for Mixxx
+// Seb Dooris, Fayaaz Ahmed, Lee Arromba
+
+function KAOSSDJ() {};
```
Please change this to `var KAOSSDJ = {};`. Declaring the variable as an empty function like that confuses people who are new to JS.

Be-ing reviewed (changes requested):
Thank you for submitting this. Overall it looks quite good. I left a few comments below.

larromba commented:
@Be-ing thanks for the feedback. all sounds good - will aim to try and get that done next week after

Be-ing commented:
We're planning to publish a release candidate for 2.1 on Tuesday April 3. Do you think you could take care of it before then?

larromba committed:
touchpad x & y axis

larromba committed:
const -> var

larromba committed:
cue_cdj -> cue_gotoandplay

larromba committed:
fix: variable declared as empty function

larromba commented:
@Be-ing added the changes

larromba commented:
@Be-ing wiki is also up to date

Be-ing commented:
Whoops, I noticed after merging that this PR was targeted at the master branch. I rebased it on 2.1, so it will be in the release.

larromba commented:
@Be-ing ok no problem. thanks!
